### ISSUE TITLE
Persistent does not implement upsert

### ISSUE THREAD

**Issue by chreekat** [yesodweb/persistent#613]
Upsert is an either/or. Either the row exists and is updated, or it does not exist and is inserted.

Persistent, however, implements a function that inserts _and then updates_ a row that does not exist. The tests expect that behavior, even though it is incorrect. What's more, Persistent's version of upsert _replaces_ a row that exists, rather than just updating it![*] Nowhere have I read that to be the desired behavior.

[*] An existing row is only replaced if upsert's second argument is `[]`, but I still think that's a bug.

In short:

|  | **Expected** | **Actual** |
| --- | --- | --- |
| **Row Exists** | updated | replaced if updates == `[]` |
| **Row Doesn't Exist** | inserted | inserted and then updated |

Correct me if I'm wrong. I'll be sending along a patch soon in the meanwhile.

---

> Upsert: An operation that inserts rows into a database table if they do not already exist, or updates them if they do.
> — https://en.wiktionary.org/wiki/upsert

\

> A relational database management system uses SQL MERGE (also called upsert) statements to INSERT new records or UPDATE existing records depending on whether [the given] condition matches.
> — https://en.wikipedia.org/wiki/Merge_(SQL)

\

> "UPSERT" is a DBMS feature that allows a DML statement's author to atomically either insert a row, or on the basis of the row already existing, UPDATE that existing row instead....
> — https://wiki.postgresql.org/wiki/UPSERT

\

Illustrative (questionable) test:

> ``` haskell
> it "upsert with updates" $ db $ do
>     deleteWhere ([] :: [Filter Upsert])
>     let email = "dude@example.com"
>     Nothing :: Maybe (Entity Upsert) <- getBy $ UniqueUpsert email
>     let up0 = Upsert email 0
>     Entity _ up1 <- upsert up0 [UpsertCounter +=. 1]
>     upsertCounter up1 @== 1
> ```
> 
> — https://github.com/yesodweb/persistent/blob/master/persistent-test/src/PersistentTest.hs#L509

`upsertCounter up1` should be 0.

**Comment by psibi** [comment]
Hey, can you show an example of this by a small snippet in both scenarios ? Does this affect the end result now ?

**Comment by chreekat** [comment]
Here's a complete example. This contrivance has users and their (unique) accounts.

I want to add two units to a user's account. If the account exists, increment the balance. If it does not, initialize it with two units.

```haskell
{-# LANGUAGE EmptyDataDecls             #-}
{-# LANGUAGE FlexibleContexts           #-}
{-# LANGUAGE GADTs                      #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE MultiParamTypeClasses      #-}
{-# LANGUAGE OverloadedStrings          #-}
{-# LANGUAGE QuasiQuotes                #-}
{-# LANGUAGE TemplateHaskell            #-}
{-# LANGUAGE TypeFamilies               #-}
import           Control.Monad.IO.Class  (liftIO)
import           Database.Persist
import           Database.Persist.Sqlite
import           Database.Persist.TH

share [mkPersist sqlSettings, mkMigrate "migrateAll"] [persistLowerCase|
Usr
Account
    usr UsrId
    balance Int
    UniqueAccount usr
    deriving Show
|]

main :: IO ()
main = runSqlite ":memory:" $ do
    runMigration migrateAll

    u <- insert Usr
    upsert (Account u 2) [AccountBalance +=. 2]
    liftIO . print =<<  map (account . entityVal) <$> selectList [] []
  where
    account :: Account -> Account
    account = id
```

On master, this prints

    [Account {accountUsr = UsrKey {unUsrKey = SqlBackendKey {unSqlBackendKey = 1}}, accountBalance = 4}]

The user's balance is 4 instead of 2.

This *definitely* affects the end result. After my patch, the balance is 2, as expected.

**Comment by psibi** [comment]
Oh yeah, this seems wrong. It *should* be `2`.

**Comment by chreekat** [comment]
Good, I'm not crazy. :)

In the new tests in #614, I also demonstrate the other bug, which is that an existing row is first *replaced*, and *then* updated. Clearly it should only be updated.

**Comment by psibi** [comment]
@chreekat There are lots of tests there in that PR and I'm lazy. Can you point out the line number in that PR (or demo it using a small snippet ) demonstrating the other "bug" which you are talking about ?

I also checked your above script for Postgres 9.5+ and it also had the same bug. Note that postgres 9.5 + has a different code flow for the upsert query and right now it's not being tested in Travis CI.

**Comment by chreekat** [comment]
Combining the tests to be efficient with boilerplate:

```haskell
{-# LANGUAGE EmptyDataDecls             #-}
{-# LANGUAGE FlexibleContexts           #-}
{-# LANGUAGE GADTs                      #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE MultiParamTypeClasses      #-}
{-# LANGUAGE OverloadedStrings          #-}
{-# LANGUAGE QuasiQuotes                #-}
{-# LANGUAGE TemplateHaskell            #-}
{-# LANGUAGE TypeFamilies               #-}
import           Control.Monad.IO.Class  (liftIO)
import           Database.Persist
import           Database.Persist.Sqlite
import           Database.Persist.TH

share [mkPersist sqlSettings, mkMigrate "migrateAll"] [persistLowerCase|
Usr
Account
    usr UsrId
    balance Int
    extra String
    UniqueAccount usr
    deriving Show
|]

main :: IO ()
main = insertBug >> updateBug

insertBug = runSqlite ":memory:" $ do
    runMigration migrateAll

    u <- insert Usr

    upsert (Account u 2 "") [AccountBalance +=. 2]
    liftIO (putStr "Should be 2: ")
    liftIO . print =<<  map (accountBalance . entityVal) <$> selectList [] []

updateBug = runSqlite ":memory:" $ do
    runMigration migrateAll

    u <- insert Usr

    insert (Account u 0 "inserted")
    upsert (Account u 100 "new value") []
    liftIO (putStr "Should be 'inserted': ")
    liftIO . print =<<  map (accountExtra . entityVal) <$> selectList [] []
```

The second bug is not as general as I first reported, but it's still a bug. I'll rewrite the issue. The second bug is specifically "If no updates are provided, an existing row is replaced." I still don't think that's what upsert is supposed to do.

**Comment by chreekat** [comment]
@psibi oh, the new backend-specific upsert is also implemented incorrectly for Postgres? I'll inspect that, but I can't test it locally atm.

**Comment by chreekat** [comment]
Yep, https://github.com/yesodweb/persistent/blob/master/persistent/Database/Persist/Sql/Orphan/PersistUnique.hs#L34

That implements the second bug by using `defaultUpsert` when updates == `[]`. I don't think the new `connUpsertSql` has the right shape for a quick fix, though. Postgres will need `ON CONFLICT (...) DO UPDATE SET (...)` when `updates /= []`, and `ON CONFLICT (...) DO NOTHING` when `updates == []`.

That's at https://github.com/yesodweb/persistent/blob/master/persistent-postgresql/Database/Persist/Postgresql.hs#L246

I can't fix this now, since I can't even use latest persistent because of esqueleto's foot-dragging. :( (I discovered this late yesterday and had to patch lts-6.17's version of persistent for my use...)

**Comment by psibi** [comment]
@chreekat Don't worry, I will fix that postgres bug as soon as I get some time. Regarding your second "update bug", I don't understand why it is a bug. Why do you think it should be `inserted` and not `new value` ? My understand is that if there is already a row present for that unique constraint, we just replace all other values for that unique constraint. Note that the `upsert` function itself will throw an exception if there are more than one unique constraint for a table.

**Comment by chreekat** [comment]
I interpret `upsert newVal []` to mean, "Insert newVal if it violates no constraints. Otherwise, do nothing."

That would be equivalent to Postgres' `insert into ... on conflict (...) do nothing`.

It would also be equivalent to MySQL's `insert... on duplicate key update`.

The behavior you are talking about is equivalent to MySQL's `replace`. I think it's a useful function, but it's not `upsert`.

**Comment by chreekat** [comment]
Another way to look at it: if I really want to replace an existing entry with `upsert`, I would write

```haskell
upsert newVal
       [Field1 =. field1 newVal, Field2 =. field2 newVal, ..., FieldN =. fieldN newVal]
```

Making `upsert newVal []` a special case that does the same thing is surprising to me. And, how then should I do the equivalent of `insert ... on conflict do nothing`?

**Comment by psibi** [comment]
Okay, I see your point. IMO, this needs a good documentation with examples on haddocks.

**Comment by psibi** [comment]
@chreekat By the way, the postgres 9.5+ server - doesn't have the "insert bug:. Yesterday, I was running an old version of the persistent library. But, I did encounter a new bug - while testing it (which is completely unrelated to this). I will send a PR for that soon.

**Comment by chreekat** [comment]
Agreed about haddocks. A version bump too, since it's changing behavior. I'm busy today, and maybe all next week, but I'll get to it if I can.

**Comment by psibi** [comment]
I'm working on a related code in Persistent and will try to fix this by tonight.

**Comment by psibi** [comment]
@chreekat This commit https://github.com/yesodweb/persistent/pull/618/commits/64a9c9aa491ccfa7279934ee4c5762f418987d4b specifically fixes both the bug described by you in the issue. Would love to hear your feedback. I will add Haddock docs soon with examples for `upsert`.

**Comment by chreekat** [comment]
Well, #614 fixes this bug as well as the tests. Did I miss something?

**Comment by mitchellwrosen** [comment]
Output of `stack script --resolver lts-8.6 bug.hs` (persistent-2.6.1) on @chreekat's first example given above:

```
[Account {accountUsr = UsrKey {unUsrKey = SqlBackendKey {unSqlBackendKey = 1}}, accountBalance = 4}]
```

Looks like the bug is still there, since `accountBalance` should be 2.

**Comment by psibi** [comment]
@mitchellwrosen The combined PR of mine and @chreekat 's is available here which fixes the problem: https://github.com/yesodweb/persistent/pull/618

Greg wanted some changes for MongoDB in that pull request to which I haven't yet gotten back to. Also, I there there are some conflicts in that file (which should be easily resolved) right now. I will see if I can give some time to the PR this weekend. If you want to take up the task of fixing the PR, feel free to do so and let me know.

**Comment by mitchellwrosen** [comment]
@psibi I _hope_ I don't have to fix up this PR, but I do need this fix in. Would be great if you or someone else with better knowledge of the code could close this one out. Thanks!

**Comment by psibi** [comment]
@mitchellwrosen I have updated the PR. Can you see if it works for you ?

**Comment by mitchellwrosen** [comment]
Both bugs originally reported by @chreekat were fixed in the latest #618. Thank you!

**Comment by psibi** [comment]
`persistent-2.7.0` has been released which fixes this. Closing the issue.
